### PR TITLE
Handle unbalanced tags

### PR DESCRIPTION
**Pull request by adon-at-work**
the current version does not purify tags that can possibly close tags of those trusted code.
`purify("</a></b>")` now leave everything untouched, but there will have undesired effect when it's appended into trusted code.

Reference: https://github.com/yahoo/html-purify/pull/7/files#diff-cfdb6a0bb85a9c9dcb6708d2225134cfR35

### CODE REVIEW

**Comment by adon-at-work** [comment]
open to discussion: whether we’d like to include a hardcoded list of tags that do not require a balanced closing tag (e.g., <input>)

**Comment by adon-at-work** [comment]
@maditya @neraliu , please help review and merge

**Comment by yahoocla** [comment]
CLA is valid!

**Comment by adon-at-work** [comment]
> the above logic still cannot handle the <!-- as it is not the tag open name state.

correct, better do it in a separate PR :)

> I am not sure whether this is the best approach to push/pop the tag, as there is a https://github.com/yahoo/html-purify/blob/master/src/derived-states.js to specify what action needs to be done in different state,

Do you mean we should update the comment there? 

>  the current implementation is for attribute value clean up

?? 
the test case has shown that it can balance tags. https://github.com/yahoo/html-purify/commit/293a26aec78ce36c831b76d29dc9bccbdb13d11c#diff-cfdb6a0bb85a9c9dcb6708d2225134cfR35.

**Comment by adon-at-work** [comment]
Please refer to https://github.com/yahoo/html-purify/commit/293a26aec78ce36c831b76d29dc9bccbdb13d11c#diff-cfdb6a0bb85a9c9dcb6708d2225134cfR35 for a detailed use case.

File: src/html-purify.js [hunk 1]
```diff
@@ -10,19 +10,21 @@ See the accompanying LICENSE file for terms.
...
             enableVoidingIEConditionalComments: true
-        };
-        this.parser = new Parser(config);
-        this.output = '';
-        this.attrVals = {};
-        this.hasSelfClosing = 0;
-        this.cssParser = new CssParser({"ver": "strict", "throwError": false});
+        }).on('postWalk', function (lastState, state, i, endsWithEOF) {
+            processTransition.call(that, lastState, state, i);
+        });
```
Comment (adon-at-work) [inline src/html-purify.js:33]: FYI, postWalk needs to be registered only once. so moved from `purify()` to here

**Comment by maditya** [comment]
I think we should make tag balancing configurable.

File: src/tag-attr-list.js [hunk 2]
```diff
@@ -239,4 +239,8 @@ Whitelist.HrefAttributes = [
     "xlink:href"
 ];
 
+// Void elements only have a start tag; end tags must not be specified for void elements.
+// https://html.spec.whatwg.org/multipage/syntax.html#void-elements
+Whitelist.VoidElements = {"area":1, "base":1, "br":1, "col":1, "embed":1, "hr":1, "img":1, "input":1, "keygen":1, "link":1, "menuitem":1, "meta":1, "param":1, "source":1, "track":1, "wbr":1};
```
Comment (maditya) [inline src/tag-attr-list.js:244]: is this faster/better for lookups, as compared to arrays we use elsewhere in this file? If so, we should probably change others to be objects like this one

File: src/html-purify.js (hunk 1 above)
Comment (maditya) [inline src/html-purify.js:33]: :+1:

File: src/html-purify.js [hunk 3]
```diff
@@ -39,51 +41,68 @@ See the accompanying LICENSE file for terms.
...
+                    if (openedTag) {
+                        this.output += '</' + openedTag + '>';
                     }
                 }
-                else if (key === "style") {// TODO: move style to a const
-                    if (this.cssParser.parseCssString(this.attrVals[key])) {
-                         attrValString += " " + key + "=" + "\"" + this.attrVals[key] + "\"";
+                else {
+                    //  - void elements only have a start tag; end tags must not be specified for void elements.
+                    this.hasSelfClosing = this.hasSelfClosing || voidElements[tagName];
```
Comment (maditya) [inline src/html-purify.js:83]: any specific reason why we're always adding a solidus for void elements? spec just says it is okay to have a solidus

File: src/tag-attr-list.js (hunk 2 above)
Comment (adon-at-work) [inline src/tag-attr-list.js:244]: object lookup is generally faster than Array.indexOf(). 
but the actual performance difference could depend on the JS engine (nodejs, Chrome, or IE...)

File: src/html-purify.js (hunk 3 above)
Comment (adon-at-work) [inline src/html-purify.js:83]: True. Not really a must, and it's optional in HTML 5. 
It seems more balanced, more defensive, is more compatible with different standards (e.g., XHTML).
What's your preference? keep or change? :)

File: src/tag-attr-list.js (hunk 2 above)
Comment (neraliu) [inline src/tag-attr-list.js:244]: r u sure that object lookup is faster?

Comment (maditya) [inline src/tag-attr-list.js:244]: cool, we'll convert other array elements to objects in another PR

**Comment by adon-at-work** [comment]
And here's an example that can help you make a holistic decision:
given `<embed>123</embed>`

first of all, embed is a void element
- with tag balancing and mandatory solidus, the output is `<embed />123`
- with tag balancing only, the output is `<embed>123`
- original implementation, the output is `<embed>123</embed>`

File: src/html-purify.js (hunk 3 above)
Comment (adon-at-work) [inline src/html-purify.js:83]: and imagine if there's an old browser not knowing a specific tag belongs to the void element, as defined in HTML 5.

given `<tag>123</tag>abc`, where tag is a void element
the old browser will understand it really as `<tag>123</tag>abc`, whereas HTML 5 compliant browser will treat it as `<tag>123abc`. expect a different DOM will be resulted.

with solidus always there, we can get `<tag />123abc`. all browsers are expected to parse it consistently. 

p.s. thinking abt whether we'd move it to the canonicalization layer. @neraliu , thoughts?

File: src/tag-attr-list.js (hunk 2 above)
Comment (adon-at-work) [inline src/tag-attr-list.js:244]: @neraliu, see http://jsfiddle.net/agup/Y8SBL/10/

**Comment by adon-at-work** [comment]
> I think we should make tag balancing configurable.

i have updated this PR to make it configurable.

File: src/html-purify.js (hunk 3 above)
Comment (maditya) [inline src/html-purify.js:83]: okay, makes sense!
+1, to move it to canonicalization layer

**Comment by maditya** [comment]
other than the merge conflict, this looks good to merge.

**Comment by adon-at-work** [comment]
@maditya merge conflict resolved. thx

**Comment by maditya** [comment]
:+1:

**Pull request merged**